packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts

```diff
@@ -0,0 +1,135 @@
+import { nls } from '../messages';
+import { notificationService } from '../notifications';
+import { sfdxCoreSettings } from '../settings';
+
+import * as vscode from 'vscode';
+
+interface SfdxProjectJson {
+  packageDirectories: [PackageDirectory];
+}
+
+interface PackageDirectory {
+  path: string;
+  default?: boolean;
+}
+
+enum FileEventType {
+  Create,
+  Change,
+  Delete
+}
+
+const WAIT_TIME_IN_MS = 1000;
+
+export async function registerPushOrDeployOnSave() {
+  if (sfdxCoreSettings.getPushOrDeployOnSaveEnabled()) {
+    const sourceFileWatcher = await createSourceFileWatcher();
+    if (sourceFileWatcher) {
+      setupFileCreateListener(sourceFileWatcher);
+      setupFileChangeListener(sourceFileWatcher);
+      setupFileDeleteListener(sourceFileWatcher);
+    }
+  }
+}
+
+function setupFileCreateListener(sourceFileWatcher: vscode.FileSystemWatcher) {
+  const createdFiles: vscode.Uri[] = [];
+  let createdFilesTimeout: NodeJS.Timer;
+  sourceFileWatcher.onDidCreate(async uri => {
+    createdFiles.push(uri);
+    clearTimeout(createdFilesTimeout);
+
+    createdFilesTimeout = setTimeout(async () => {
+      await pushOrDeploy(FileEventType.Create, createdFiles);
+    }, WAIT_TIME_IN_MS);
+  });
+}
+
+function setupFileChangeListener(sourceFileWatcher: vscode.FileSystemWatcher) {
+  sourceFileWatcher.onDidChange(async uri => {
+    await pushOrDeploy(FileEventType.Change, [uri]);
+  });
+}
+
+function setupFileDeleteListener(sourceFileWatcher: vscode.FileSystemWatcher) {
+  let deletedFilesTimeout: NodeJS.Timer;
+  sourceFileWatcher.onDidDelete(async uri => {
+    clearTimeout(deletedFilesTimeout);
+
+    deletedFilesTimeout = setTimeout(async () => {
+      await pushOrDeploy(FileEventType.Delete);
+    }, WAIT_TIME_IN_MS);
+  });
+}
+
+async function pushOrDeploy(
+  fileEventType: FileEventType,
+  filesToDeploy?: vscode.Uri[]
+): Promise<void> {
+  try {
+    const orgType = await getWorkspaceOrgType();
+    if (orgType === OrgType.SourceTracked) {
+      vscode.commands.executeCommand('sfdx.force.source.push');
+    }
+
+    if (orgType === OrgType.NonSourceTracked) {
+      if (fileEventType === FileEventType.Create) {
+        vscode.commands.executeCommand(
+          'sfdx.force.source.deploy.multiple.source.paths',
+          filesToDeploy!.slice(0)
+        );
+      } else if (fileEventType === FileEventType.Change) {
+        vscode.commands.executeCommand(
+          'sfdx.force.source.deploy.manifest.or.source.path',
+          filesToDeploy![0]
+        );
+      } else if (fileEventType === FileEventType.Delete) {
+        notificationService.showErrorMessage(
+          nls.localize('error_change_not_deleted_text')
+        );
+      }
+    }
+  } catch (e) {
+    if (e.name === 'NamedOrgNotFound') {
+      notificationService.showErrorMessage(
+        nls.localize('error_fetching_auth_info_text')
+      );
+    } else if (e.name === 'NoDefaultusernameSet') {
+      // Do nothing.
```

Comment:
@ntotten, if no defaultusername is set in the workspace, should I notify the customer that their changes are not being deployed/pushed to any org?

Reply:
I think it would be a good idea. Maybe something in the console? "Deploy on save is enabled, but no org is authorized. No files deployed to org." //cc @ruthemmanuelle